src/Oscoin/P2P/Disco/Trace.hs

```diff
@@ -0,0 +1,22 @@
+module Oscoin.P2P.Disco.Trace
+    ( DiscoEvent (..)
```

Comment:
Nice -- Something else you could do (but I am not sure if circular deps will bite you) is to have a module like `Oscoin.Telemetry.Events.Disco` and put the data type there.

Having said that, I think I am going to steal your pattern here, as I wanted to do the same for syncing events but it didn't occur to me to put the data type in the `Oscoin.Protocol.Sync` namespace, but rather I did create a `Oscoin.Telemetry.Events.Sync`, with the result I had to use a totally opaque stringly-typed exception type as I did hit circular deps 😬

Reply:
It is indeed necessary to split into many modules due to circular dependencies -- but there's also a modularity argument: the _application_ code defines what events it exposes, and `Oscoin.Telemetry` merely interprets them. If you would think of this namespace as a proper package, with your approach _everything_ would depend on `telemetry`.